File: arrays_hashing/groupAnagrams.py
class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        anaMap = {}
        anaList = []
        for i,val in enumerate(strs):
            sortedWord = list(val)
            sortedWord.sort()
            sortedWord = "".join(sortedWord)


            if sortedWord not in anaMap:
                anaMap[sortedWord] = [i]
            else:
                anaMap[sortedWord].append(i)
        
        for key,valList in anaMap.items():
            row = [strs[i] for i in valList]
            anaList.append(row)

        return anaList
    
def main():
    s = Solution()
    print(s.groupAnagrams(["eat","tea","tan","ate","nat","bat"]))
    print(s.groupAnagrams([""]))
    print(s.groupAnagrams(["a"]))

if __name__ == "__main__":
    main()

File: arrays_hashing/groupAnagrams.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map <string,vector<int>> anaMap;
        vector<vector<string>> anaList;
        
        string curString;
        for (int i=0; i<strs.size(); i++){
            curString = strs[i];
            std::sort(curString.begin(),curString.end());
            anaMap[curString].push_back(i);
        }

        for (const auto& [key,val] : anaMap){
            vector<string> row;
            for (const auto& index : val){
                row.push_back(strs[index]);
            }
            anaList.push_back(row);
        }
        return anaList;
    }
};